sonic_fan/fan_base.py

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env python
+#
+# fan_base.py
+#
+# Base class for implementing platform-specific
+#  FAN control functionality for SONiC
+#
+
+try:
+    import abc
+except ImportError as e:
+    raise ImportError (str(e) + " - required module not found")
+
+class FanBase(object):
+    __metaclass__ = abc.ABCMeta
+
+    def get_num_fans(self):
+        """
+        Retrieves the number of FANs supported on the device
+
+        :return: An integer, the number of FANs supported on the device
+        """
+        return 0
+
+    def get_status(self, index):
+        """
+        Retrieves the operational status of FAN defined
+                by index 1-based <index>
+
+        :param index: An integer, 1-based index of the PSU of which to query status
+        :return: Boolean,
+            - True if FAN is running with some speed 
+            - False if FAN has stopped running
+        """
+        return False
+
+    def get_presence(self, index):
+        """
+        Retrieves the presence status of a FAN defined
+                by 1-based index <index>
+
+        :param index: An integer, 1-based index of the FAN of which to query status
+        :return: Boolean, True if FAN is plugged, False if not
+        """
+        return False
+    
+    def get_direction(self, index):
+        """
+        Retrieves the airflow direction of a FAN defined
+                by 1-based index <index>
+
+        :param index: An integer, 1-based index of the FAN of which to query status
+        :return: string, denoting FAN airflow direction
+        """
+        return ""
+    
+    def get_speed(self, index):
+        """
+        Retrieves the speed of a Front FAN in the tray in revolutions per minute defined
+                by 1-based index <index>
+
+        :param index: An integer, 1-based index of the FAN of which to query speed
+        :return: integer, denoting front FAN speed
+        """
+        return 0
+
+    def get_speed_rear(self, index):
+        """
+        Retrieves the speed of a rear FAN in the tray in revolutions per minute defined
+                by 1-based index <index>
```

Comment:
my understanding is that although there could have multi fans(usually 2) in a tray, but each fan has its own tachometer, so these fans in the same tray should still be treated as separated fans, In my view, this function is not needed.

Reply:
While mentioning the platform inventory, one fan-tray is usually considered as one fan unit.  ODM too, provide the APIs to retrieve 'rear-fan' speed. 'lm-sensors' (or sensors command) denotes rear fan of FAN tray 1 as fan11, rear fan of FAN tray 2 as fan22 etc.  

Open Config model FAN related CLIs also provide the fan speed in term of front and rear speed. 

Set speed FAN APIs also set speed of both the fans in a single call. Hence a separate API is provided for rear FAN speed.

Reply:
on some vendors platform, they do treat fans in the same tray as separated fans and hide the rear/front info from user, in this case, this function is not applicable.

Reply:
Hi Keboliu,
If some vendors treat 'rear' fans as separate, indexing them is left to discretion of the vendors. e.g Accton uses,
fan1 (front) --------- fan11 (rear)
fan2 (front) --------- fan12(rear) etc

Some other platform may use fan1 (tray1-front), fan2(tray1-rear), fan3 (tray2-front), fan4 (tray2-rear) etc.
 
If separate indexing is being used, vendors can use 'get_speed' API. If tray is numbered and fans inside a tray are mentioned as 'front' and 'rear' then let 'get_speed_rear' API be used. Since some vendors also use tray numbering, can't we keep both APIs?

Reply:
I am ok with keeping both APIs.

Reply:
I think this function may require more thorough commenting to ensure vendors understand that `get_speed_rear()` is only to be implemented  for the rear fan of a 2-fan tray.

Reply:
Added extra comment to mention the same.